airlock/business_logic.py

```diff
@@ -1539,6 +1542,108 @@ def add_file_to_request(
 
         return release_request
 
+    def update_file_in_request(
+        self,
+        release_request: ReleaseRequest,
+        relpath: UrlPath,
+        user: User,
+        group_name: str = "default",
+        filetype: RequestFileType = RequestFileType.OUTPUT,
+    ) -> ReleaseRequest:
+        self._validate_editable(release_request, user)
+
+        relpath = UrlPath(relpath)
+        if not is_valid_file_type(Path(relpath)):
+            raise self.RequestPermissionDenied(
+                f"Cannot update file of type {relpath.suffix} in request"
+            )
+
+        workspace = self.get_workspace(release_request.workspace, user)
+        if (
+            workspace.get_workspace_status(UrlPath(relpath))
+            != WorkspaceFileStatus.CONTENT_UPDATED
+        ):
+            raise self.RequestPermissionDenied(
+                "Cannot update file in request if it is not updated on disk"
+            )
+
+        src = workspace.abspath(relpath)
+        file_id = store_file(release_request, src)
+
+        manifest = workspace.get_manifest_for_file(relpath)
+        assert (
+            manifest["content_hash"] == file_id
+        ), "File hash does not match manifest.json"
+
+        reviews = release_request.get_request_file_from_output_path(relpath).reviews
```

Comment:
> I think it is necessary to reset the reviews & withdraw the file

yes, I had some discussion with @bloodearnest about this whilst working on the ticket - I think it is necessary to reset reviews & withdraw because of the foreign keys. Therefore if we build that using the current functions (by which I mean e.g. `dal.reset_review_file()`) we need audit events for everything (or allow those functions to support not creating an audit event :/ ).

One reason that I wanted to keep the reset review audit log event was because I thought that an output checker who was skimming the audit logs in order to see why a file that they had previously reviewed now didn't have a review would be more likely to notice "review reset" than "file updated". But I'm really not wedded to that idea! 

> Perhaps we need to push this down to the DAL

yes, I wondered about that before. I partly didn't because in order to do that right now, I'd need to iterate through the reviews in the BLL to create the audit events for each review/withdrawal & pass through some kind of dict of audit events to the DAL, then iterate again in the DAL. The way it is in the PR, the DAL implementation bleeds into the BLL a bit, but that would still happen with the audit logs - and I think we have to generate them in the BLL. It could be done, though, especially if we decide to drop audit events for the e.g. reset review.

> multiple calls to DAL methods that will each be within their own transaction

that would be a nice benefit!

Reply:
This is exactly the kind of issue that prompted me to add this topic for discussion in the team meeting today.

We need to figure out the transactional of the DAL API. We've attempted to keep the logic in the BLL, but that means adding a list of audits or similar. Pushing some parts of the logic down into the DAL will improve the transactionality, but the DAL will grow responsiblity and complexit.

I think we also need some discussion about what exactly needs auditing, and when, which might impact on this question.

Reply:
Given that the BLL/DAL/auditing discussion may take a while, I'm happy to +1 this as it is.

Reply:
ok, thanks, let's merge